mountpoint-s3/src/prefetch/backpressure_controller.rs

```diff
@@ -0,0 +1,192 @@
+                        read_window_end_offset = self.read_window_end_offset,
+                        to_increase,
+                        "incrementing read window"
+                    );
+                    self.increment_read_window(to_increase).await?;
+                    self.read_window_end_offset = new_read_window_end_offset;
+                }
+            }
+            BackpressureFeedbackEvent::PartQueueStall => self.try_scaling_up(),
+        }
+        Ok(())
+    }
+
+    // Send an increment read window request to the stream producer
+    async fn increment_read_window<E>(&self, len: usize) -> Result<(), PrefetchReadError<E>> {
+        // This should not block since the channel is unbounded
+        self.read_window_updater
+            .send(len)
+            .await
+            .map_err(|_| PrefetchReadError::ReadWindowIncrement)
+    }
+
+    fn remaining_window(&self) -> usize {
+        self.read_window_end_offset.saturating_sub(self.next_read_offset) as usize
+    }
+
+    // Try scaling up preferred read window size with a multiplier configured at initialization.
+    fn try_scaling_up(&mut self) {
+        if self.preferred_read_window_size < self.max_read_window_size {
+            let new_read_window_size =
+                (self.preferred_read_window_size * self.read_window_size_multiplier).min(self.max_read_window_size);
+            trace!(
+                current_size = self.preferred_read_window_size,
+                new_size = new_read_window_size,
+                "scaling up preferred read window"
+            );
+            self.preferred_read_window_size = new_read_window_size;
+        }
+    }
+}
+
+impl BackpressureLimiter {
+    pub fn read_window_end_offset(&self) -> u64 {
+        self.read_window_end_offset
+    }
+
+    /// Checks if there is enough read window to put the next item with a given offset to the stream.
+    /// It blocks until receiving enough incrementing read window requests to serve the next part.
+    ///
+    /// Returns the new read window offset.
+    pub async fn wait_for_read_window_increment<E>(
+        &mut self,
+        offset: u64,
+    ) -> Result<Option<u64>, PrefetchReadError<E>> {
+        // There is already enough read window so no need to block
+        if self.read_window_end_offset > offset {
+            // Check the read window incrementing queue to see there is an early request to increase read window
+            let new_read_window_offset = if let Ok(len) = self.read_window_incrementing_queue.try_recv() {
+                self.read_window_end_offset += len as u64;
+                Some(self.read_window_end_offset)
+            } else {
+                None
+            };
+            return Ok(new_read_window_offset);
+        }
+
+        // Reaching here means there is not enough read window, so we block until it is large enough
+        while self.read_window_end_offset <= offset && self.read_window_end_offset < self.request_end_offset {
```

Comment:
is `self.read_window_end_offset < self.request_end_offset` a precaution or before the last commit we actually could've get blocked, when `offset` was above the object length?

Reply:
It could be blocked after the request is completed, this check prevents that.

Reply:
Ah, I see, for `read_from_request` the provided `offset` is always `<= object_size`. When `offset == object_size` we now that it won't be read cause EOF will be reached, but in other cases we want window to be `self.read_window_end_offset > offset`.

[For the caching use case](https://github.com/awslabs/mountpoint-s3/blob/ade08b63e18f21759620fc8dc75cfafde702e9cf/mountpoint-s3/src/prefetch/caching_stream.rs#L140), where we wait for increment before loading data from disk, may next `block_offset` be `>= object_size`?

Reply:
Yeah, previously we could have a problem after serving the last cache block. Now we will block until `self.read_window_end_offset` has been incremented to the end of the request then we bail out of the loop with second condition (even if the first condition is still true because `block_offset` is larger than `self.read_window_end_offset`).

Reply:
It gets blocked and that was bad, but we don't see it because the reader receives all the data they requested so they close the file and everything is dropped.